R/vsc_r_api.R

```diff
@@ -0,0 +1,334 @@
+    }
+
+    invisible(
+        request_response(
+            "document_new",
+            text = text,
+            type = type,
+            position = position
+        )
+    )
+}
+
+restart_r_session <- function() {
+    invisible(
+        request_response("restart_r")
+    )
+}
+
+vsc_viewer <- function(url, height = NULL) {
+    # cant bind to this directly because it's not created when the bidning is
+    # made.
+    .vsc.viewer(url)
+}
+
+rstudio_vsc_mapping <-
+    list(
+        getActiveDocumentContext = get_active_document_context,
+        getSourceEditorContext = get_active_document_context,
+        isAvailable = is_available,
+        verifyAvailable = verify_available,
+        insertText = insert_or_modify_text,
+        modifyRange = insert_or_modify_text,
+        readPreference = read_preference,
+        readRStudioPreference = read_preference,
+        hasFun = has_fun,
+        findFun = get_fun,
+        showDialog = show_dialog,
+        navigateToFile = navigate_to_file,
+        setSelectionRanges = set_selection_ranges,
+        setCursorPosition = set_cursor_position,
+        documentSave = document_save,
+        documentId = document_id,
+        documentPath = document_path,
+        documentSaveAll = document_save_all,
+        documentNew = document_new,
+        getActiveProject = get_active_project,
+        restartSession = restart_r_session,
+        viewer = vsc_viewer
+    )
+
+rstudio_vsc_no_map <-
+    list(
+        "getConsoleEditorContext",
+        "sourceMarkers",
+        "versionInfo",
+        "getVersion",
+        "documentClose",
+        "sendToConsole",
+        "showPrompt",
+        "showQuestion",
+        "updateDialog",
+        "openProject",
+        "initializeProject",
+        "addTheme",
+        "applyTheme",
+        "convertTheme",
+        "getThemeInfo",
+        "getThemes",
+        "removeTheme",
+        "jobAdd",
+        "jobAddOutput",
+        "jobAddProgress",
+        "jobRemove",
+        "jobRunScript",
+        "jobSetProgress",
+        "jobSetState",
+        "jobSetStatus",
+        "launcherGetInfo",
+        "launcherAvailable",
+        "launcherGetJobs",
+        "launcherConfig",
+        "launcherContainer",
+        "launcherControlJob",
+        "launcherGetJob",
+        "launcherHostMount",
+        "launcherNfsMount",
+        "launcherPlacementConstraint",
+        "launcherResourceLimit",
+        "launcherSubmitJob",
+        "launcherSubmitR",
+        "previewRd",
+        "previewSql",
+        "writePreference",
+        "writeRStudioPreference",
+        "getPersistentValue",
+        "setPersistentValue",
+        "savePlotAsImage",
+        "createProjectTemplate",
+        "hasColourConsole",
+        "bugReport",
+        "buildToolsCheck",
+        "buildToolsInstall",
+        "buildToolsExec",
+        "dictionariesPath",
+        "userDictionariesPath",
+        "executeCommand",
+        "translateLocalUrl"
+    )
```

Comment:
The mapping and no mapping lists look a bit redundant which duplicate names with different naming conventions.

Does it look better if we dedicate a file in which the original names are used to define mapped functions, and we source the file to an environment on session attach, and replace all functions that exist in both this mapping environment and `rstudioapi` namespace with original names. Then there's no need to maintain such mapping lists.

For example, a dedicated file could look like

```r
getActiveDocumentContext <- function(...) ...
getSourceEditorContext <- function(...) ...

sourceMarkers <- not_yet_implemented
```

Reply:
This could work although things get a bit tricky to make robust.

* I am not replacing all functions in the `rstudioapi` namespace since there are some that remain useful, like those that deal with constructing/testing positions and ranges. 
* The functions that are to be assigned into the namespace currently share an environment with a bunch of helper functions, those shouldn't be assigned into the `rstudioapi` namespace - most likely they wouldn't because they don't match with a function in the namespace, but there is a small chance they could? I certainly haven't checked all unexported methods.
* So to be robust to collision, we would have to prep two environments: A child environment to contain the methods to be iterated over for assignment into API namespace, and a parent environment, which is not iterated over, which contains the helper methods those API replacements need to function.

I didn't think too hard about it, but having the mappings seemed a nice simple way to ensure robustness.

I thought snake case for the VSC-R api was convenient, since although some functions are a 1:1 mapping with the `rstudioapi`, others are not, and it would save some confusion.

Reply:
I guess an advantage of your proprosal would be the the api functions do not fill up the `.vsc` environment, making that a bit easier to use by inspection. Not sure how important that is?

Reply:
Okay I've refactored along these lines in cd1fb08.

This looks a little spooky to me: https://github.com/Ikuyadeu/vscode-R/pull/408/files#diff-2711d8988b4edf6562a0668d7e873c13R435

But it gets the job done.

Reply:
There are now only two potential collisions:

* .vsc_not_yet_implemented
* .vsc_document_context

and those prefixes should be good enough to guard against that?